src/__spec__/interleave-generator.spec.ts

```diff
@@ -4,127 +4,83 @@ import * as iter from '../index'
 assert<
   IterableIterator<string>
 >(
-  iter.interleaveGenerator(
-    function * (
-      canTakeAny: Function,
-      buffer: iter.InterleaveBuffer<string>
+  iter.interleaveGenerator<string>(
+    function * gen (
+      canTakeAny,
+      buffer
     ) {
-      yield buffer.take()
+      assert<iter.InterleaveCallback<iter.InterleaveBuffer<string>[]>>(gen)
+      assert<iter.InterleaveBuffer<string>>(buffer)
+      assert<iter.InterleaveBuffer<string> | null>(canTakeAny())
+      assert<string | undefined>(buffer.take())
+      assert<boolean>(buffer.canTake())
+
+      if (buffer.canTake()) {
+        assert<iter.TakableInterleaveBuffer<string>>(buffer)
+        const res = buffer.take()
+        assert<string>(res)
+        yield res
+      }
     }
   )(['foo'])
 )
 
 assert<
-  IterableIterator<string | number>
+  IterableIterator<string>
 >(
-  iter.interleaveGenerator(
-    function * (
-      canTakeAny: Function,
-      b1: iter.InterleaveBuffer<string>,
-      b2: iter.InterleaveBuffer<number>
+  iter.interleaveGenerator<string>(
+    function * gen (
+      canTakeAny,
+      ...buffers
     ) {
-      yield b1.take()
-      yield b2.take()
-    }
-  )(['foo'], [2])
-)
+      assert<iter.InterleaveCallback<iter.InterleaveBuffer<string>[]>>(gen)
+      assert<iter.InterleaveBuffer<string>[]>(buffers)
+      assert<iter.InterleaveBuffer<string> | null>(canTakeAny())
 
-assert<
-  IterableIterator<string | number | Function>
->(
-  iter.interleaveGenerator(
-    function * (
-      canTakeAny: Function,
-      b1: iter.InterleaveBuffer<string>,
-      b2: iter.InterleaveBuffer<number>,
-      b3: iter.InterleaveBuffer<Function>
-    ) {
-      yield b1.take()
-      yield b2.take()
-      yield b3.take()
-    }
-  )(['foo'], [2], [(_: any) => _])
-)
+      for (const buffer of buffers) {
+        assert<string | undefined>(buffer.take())
+        assert<boolean>(buffer.canTake())
 
-assert<
```

Comment:
You can't just remove these. Sad though it is, when we have to make five overloads of the type to account for situations with different parameters, we need five tests, one for each variant. I think it's important to keep these tests because it can be particularly hard to spot subtle differences in those highly duplicated type overloads.

Reply:
Your criticism is valid. However, as I said in the original comment, this is only temporary, I plan to replace the many overloads with a single one that covers all use cases.

Reply:
How is that possible?

Reply:
Everything seems impossible until we actually try.

Reply:
Hahaha OK then. I'm just saying that all the type code I've seen uses multiple overloads to avoid a design limitation of typescript: that its spread type is unable to infer a tuple type, forcing instead an array like `Array<string|number>`. While that's not exactly wrong as a type it allows you to put the parameters in the wrong positions. I don't see how you'll get around this, but I'll be pretty impressed if you manage it.

Reply:
I can manage this, but there's a catch: User must always specify generic parameters, like this:

```
const generator = interleaveGenerator<[string, number, boolean]> (function * (
  canTakeAny,
  someString,
  someNumber,
  someBoolean
) { /* ... */ })
```

I'll do it if you're OK with this.

Reply:
Unless I'm much mistaken the types I wrote allow you to do that without any changes at all, don't they? And the user doesn't always have to specify the generic parameters: if they prefer they can eithe import InterleaveBuffer and specify the types of the parameters to the function in terms of `InterleaveBuffer<T>`.

If I just make my `T_` types default to `any` the really ugly error messages go away too when I forget to specify types using either method, and I'm just left with standard warnings about having a function with untyped parameters.

Let me draw up those changes and improve the type tests and see what you think.

Reply:
In my current implementation, user doesn't always have to specify generic parameters, but they must specify type annotations of callback function.